File: mts-asn1-kotlin-per/src/test/java/com/ericsson/mts/asn1/AbstractTests.java
// SPDX-License-Identifier: MIT
/*
 * Copyright 2019 Ericsson, https://www.ericsson.com/en
 * Copyright 2024 Andrea Mennillo, https://amennillo.eu
 */
package com.ericsson.mts.asn1;

import org.apache.commons.io.IOUtils;
import org.custommonkey.xmlunit.Diff;
import org.custommonkey.xmlunit.XMLUnit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.File;
import java.io.InputStream;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.nio.charset.StandardCharsets;

import static net.javacrumbs.jsonunit.JsonAssert.assertJsonEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class AbstractTests {
  static ASN1Translator asn1Translator;
  protected Logger logger = LoggerFactory.getLogger(this.getClass().getSimpleName());

  void test(String type, String binaryPath, String expectedJsonPath, String expectedXmlPath) throws Exception {
    testEncode(type, binaryPath, expectedJsonPath, expectedXmlPath);
    testDecode(type, binaryPath, expectedJsonPath, expectedXmlPath);

//        updateDataformatFileTest(type, binaryPath, expectedJsonPath, expectedXmlPath);
  }

  protected void testEncode(String type, String binaryPath, String expectedJsonPath, String expectedXmlPath) throws Exception {
    //JSON encode test
    {

      KotlinJsonFormatReader jsonFormatReader = new KotlinJsonFormatReader(this.getClass().getResourceAsStream(expectedJsonPath), type);
      BitArray bitArray = new BitArray();
      asn1Translator.encode(type, bitArray, jsonFormatReader);


      InputStream expectedInputStream = this.getClass().getResourceAsStream(binaryPath);
      BitArray bitArray1 = new BitArray();
      while (expectedInputStream.available() != 0) {
        bitArray1.write(expectedInputStream.read());
      }

      assertEquals(bitArray1.getBinaryMessage(), bitArray.getBinaryMessage());
    }

    //XML encode test
    {
      XMLFormatReader xmlFormatReader = new XMLFormatReader(this.getClass().getResourceAsStream(expectedXmlPath), type);
      BitArray bitArray = new BitArray();
      asn1Translator.encode(type, bitArray, xmlFormatReader);

      InputStream expectedInputStream = this.getClass().getResourceAsStream(binaryPath);
      BitArray bitArray1 = new BitArray();
      while (expectedInputStream.available() != 0) {
        bitArray1.write(expectedInputStream.read());
      }
      assertEquals(bitArray1.getBinaryMessage(), bitArray.getBinaryMessage());
    }
  }

  void testDecode(String type, String binaryPath, String expectedJsonPath, String expectedXmlPath) throws Exception {
    //JSON decode test
    {
      KotlinJsonFormatWriter formatWriter = new KotlinJsonFormatWriter();
      asn1Translator.decode(type, this.getClass().getResourceAsStream(binaryPath), formatWriter);
      String actual = formatWriter.getJsonNode().toString();
      String expected = IOUtils.toString(this.getClass().getResourceAsStream(expectedJsonPath), StandardCharsets.UTF_8);
      assertJsonEquals(expected, actual);
    }

    //XML decode test
    {
      XMLFormatWriter formatWriter = new XMLFormatWriter();
      asn1Translator.decode(type, this.getClass().getResourceAsStream(binaryPath), formatWriter);
      TransformerFactory tf = TransformerFactory.newInstance();
      Transformer transformer = tf.newTransformer();
      transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
      transformer.setOutputProperty(OutputKeys.INDENT, "yes");
      transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");
      transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4");
      StringWriter writer = new StringWriter();
      transformer.transform(new DOMSource(formatWriter.getResult()), new StreamResult(writer));


      String actual = writer.toString();
      String expected = IOUtils.toString(this.getClass().getResourceAsStream(expectedXmlPath), StandardCharsets.UTF_8);
      XMLUnit.setIgnoreWhitespace(true);
      Diff diff = XMLUnit.compareXML(expected, actual);

      assertTrue(diff.similar(), diff + "\nActual document is :\n" + actual);
    }
  }

  /**
   * Update xml and json test file in the target folder.
   * <p>
   * WARNING : Be sure that encoding and decoding are correct before using this method.
   *
   * @param type           Entry point for a given protocol (protocol dependant)
   * @param binaryPath     File binary path
   * @param targetJsonPath Target JSON file
   * @param targetXmlPath  Target XML file
   * @throws Exception Encoding exception
   */
  void updateDataformatFileTest(String type, String binaryPath, String targetJsonPath, String targetXmlPath) throws Exception {
    //JSON decode test
    {
      KotlinJsonFormatWriter formatWriter = new KotlinJsonFormatWriter();
      asn1Translator.decode(type, this.getClass().getResourceAsStream(binaryPath), formatWriter);
      PrintWriter printWriter = new PrintWriter(new File(this.getClass().getResource(targetJsonPath).getPath()));
      printWriter.print(formatWriter.getJsonNode().toString());
      printWriter.close();
    }

    //XML decode test
    {
      XMLFormatWriter formatWriter = new XMLFormatWriter();
      asn1Translator.decode(type, this.getClass().getResourceAsStream(binaryPath), formatWriter);
      TransformerFactory tf = TransformerFactory.newInstance();
      Transformer transformer = tf.newTransformer();
      transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
      transformer.setOutputProperty(OutputKeys.INDENT, "yes");
      transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");
      transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4");
      StringWriter writer = new StringWriter();
      transformer.transform(new DOMSource(formatWriter.getResult()), new StreamResult(writer));
      PrintWriter printWriter = new PrintWriter(new File(this.getClass().getResource(targetXmlPath).getPath()));
      printWriter.print(writer);
      printWriter.close();
    }
  }

  void testParsingANTLR() {
    asn1Translator.parseTranslators();
  }
}

File: mts-asn1-converter/src/test/java/com/ericsson/mts/asn1/AbstractTests.kt
/*
 * Copyright 2022 Andrea Mennillo, git at amennillo dot eu
 * Copyright 2019 Ericsson, https://www.ericsson.com/en
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.ericsson.mts.asn1

import com.fasterxml.jackson.databind.ObjectMapper
import net.javacrumbs.jsonunit.JsonAssert
import org.custommonkey.xmlunit.XMLUnit
import org.junit.jupiter.api.Assertions
import java.io.InputStream
import java.io.StringWriter
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

open class AbstractTests {
    private val mapper = ObjectMapper()
    private val writer = mapper.writer()

    @Throws(Exception::class)
    fun test(type: String, textPath: String, expectedJsonPath: String, expectedXmlPath: String) {
        testConversion(type, textPath, expectedJsonPath, expectedXmlPath)
    }

    @Throws(Exception::class)
    fun testConversion(type: String, textPath: String, expectedJsonPath: String, expectedXmlPath: String) {
        //JSON conversion test
        run {
            val formatWriter = JSONFormatWriter()
            asn1Converter?.convert(type, getResourceAsText(textPath)!!, formatWriter)
            val actual = writer.writeValueAsString(formatWriter.jsonNode)
            val expected = getResourceAsText(expectedJsonPath)!!
            JsonAssert.assertJsonEquals(expected, actual)
        }

        //XML conversion test
        run {
            val formatWriter = XMLFormatWriter()
            asn1Converter?.convert(type, getResourceAsText(textPath)!!, formatWriter)
            val tf = TransformerFactory.newInstance()
            val transformer = tf.newTransformer()
            transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes")
            transformer.setOutputProperty(OutputKeys.INDENT, "yes")
            transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8")
            transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4")
            val writer = StringWriter()
            transformer.transform(DOMSource(formatWriter.result), StreamResult(writer))
            val actual = writer.toString()
            val expected = getResourceAsText(expectedXmlPath)!!
            XMLUnit.setIgnoreWhitespace(true)
            val diff = XMLUnit.compareXML(expected, actual)
            Assertions.assertTrue(diff.similar(), "$diff\nActual document is :\n$actual")
        }
    }

    private fun getResourceAsText(path: String): String? =
        object {}.javaClass.getResource(path)?.readText()

    companion object {
        var asn1Converter: ASN1Converter? = null

        @JvmStatic
        internal fun getResourceAsStream(path: String): InputStream? =
            object {}.javaClass.getResourceAsStream(path)
    }
}